### PR TITLE
Handle large chain calc better

### PR DESCRIPTION
**Pull request by erikjohnston**
We calculate the auth chain links outside of the main persist event transaction to ensure that we do not block other event sending during the calculation.

### CODE REVIEW

**Comment by morguldir** [comment]
- Before:
`matrix-synapse-generic-worker-8fc645778-v4cv7 generic-worker 2024-06-17 16:12:35,614 - synapse.access.http.8083 - 473 - INFO - PUT-462- <ip> - 8083 - {@morguldir:sulian.eu} Processed request: 302.537sec/0.000sec (0.419sec, 0.052sec) (297.116sec/1.248sec/19) 59B 200 "PUT /_matrix/client/v3/rooms/!OGEhHVWSdvArJzumhm%3Amatrix.org/send/m.reaction/m1718640453033.4 HTTP/1.1" "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36.0 (KHTML, like Gecko)`
- After:
`matrix-synapse-generic-worker-6449f78659-jgl26 generic-worker 2024-06-17 16:23:28,971 - synapse.access.http.8083 - 473 - INFO - PUT-131- <ip> - 8083 - {@morguldir:sulian.eu} Processed request: 271.939sec/0.140sec (0.415sec, 0.078sec) (13.365sec/15.000sec/19) 59B 200 "PUT /_matrix/client/v3/rooms/!OGEhHVWSdvArJzumhm%3Amatrix.org/send/m.reaction/m1718641136803.3 HTTP/1.1" "Mozilla/5.0 (X11; Linux x86_64; rv:127.0) Gecko/20100101 Firefox/127.0" [3 dbevts]`
Chrome/124.0.6367.243 Safari/537.36.0" [3 dbevts]`
- With #17291 
`matrix-synapse-generic-worker-685b6d5b8-g9p97 generic-worker 2024-06-17 16:43:04,772 - synapse.access.http.8083 - 473 - INFO - PUT-626- <ip> - 8083 - {@morguldir:sulian.eu} Processed request: 9.251sec/0.001sec (0.276sec, 0.040sec) (11.001sec/2.227sec/20) 59B 200 "PUT /_matrix/client/v3/rooms/!OGEhHVWSdvArJzumhm%3Amatrix.org/send/m.reaction/m1718642575413.4 HTTP/1.1" "Mozilla/5.0 (X11; Linux x86_64; rv:127.0) Gecko/20100101 Firefox/127.0" [3 dbevts]`

So this doesn't seem to have helped too much in this case, example is from HQ

I keep seeing this in my logs as well, but not sure if it's relevant
`matrix-synapse-generic-worker-685b6d5b8-g9p97 generic-worker 2024-06-17 16:42:57,777 - synapse.storage.databases.main.event_federation - 274 - INFO - GET-19- Unexpectedly found that events don't have chain IDs in room !OGEhHVWSdvArJzumhm:matrix.org: {'$e3j8FpK-lcibZO57q50gANeLSK0dptpF6fCUmBF3rKU', '$7_ZXqJstZZrXU1S-2jgKaEOpgrvWEXdU_NxQ6mxC4_8', '$QDzjkSbQgHTZPi8G2ZnzQ6TpUVmzx1Xf7wmbmWbUqWM'}
`

**Comment by erikjohnston** [comment]
Hi @morguldir, yup, this PR won't cause the event itself to get persisted faster. Instead it ensures other events (in other rooms) don't get blocked from being sent down to the client.

File: synapse/storage/databases/main/events.py [hunk 1]
```diff
@@ -100,6 +99,14 @@ def is_noop(self) -> bool:
         return not self.to_delete and not self.to_insert and not self.no_longer_in_room
 
 
+@attr.s(slots=True, auto_attribs=True)
+class NewEventChainLinks:
```
Comment (anoadragon453) [inline synapse/storage/databases/main/events.py:103]: Could you add a docstring to this object? I don't think it's particularly grokkable at a glance without going and seeing how it's used.

**Review by anoadragon453 — APPROVED** [review]
Just a couple documentation changes, otherwise LGTM!

File: synapse/storage/databases/main/events.py [hunk 2]
```diff
@@ -242,6 +251,64 @@ async def _persist_events_and_state_updates(
...
+            table="rooms",
+            keyvalues={"room_id": room_id},
+            retcols=("room_id", "has_auth_chain_index"),
+            allow_none=True,
+        )
+        if row is None:
+            return {}
+
+        if not state_events:
+            return {}
```
[the remark this answers is not in the document]
Comment (erikjohnston) [reply synapse/storage/databases/main/events.py:316]: We already do this before we start the transaction. This check is post filtering out already persisted events

File: synapse/storage/databases/main/events.py (hunk 1 above)
Comment (anoadragon453) [reply synapse/storage/databases/main/events.py:103]: I had to re-read a couple times, but it's understandable :+1: 

It's certainly a fairly abstract concept to describe without a diagram.

**Review by anoadragon453 — APPROVED** [review]

**Pull request merged**